src/Conversion/ONNXToMhlo/Math/MatMul.cpp

```diff
@@ -0,0 +1,123 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===----------------- Matmul.cpp - Lowering Matmul Op --------------------===//
+//
+// Copyright 2019-2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX Matmul Operator to Mhlo dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToMhlo/ONNXToMhloCommon.hpp"
+#include "src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp"
+#include "src/Support/TypeUtilities.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+struct ONNXMatMulOpLoweringToMhlo : public ConversionPattern {
+  ONNXMatMulOpLoweringToMhlo(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXMatMulOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXMatMulOpAdaptor operandAdaptor(operands);
+    ONNXMatMulOp matMulOp = llvm::cast<ONNXMatMulOp>(op);
+    Location loc = op->getLoc();
+    ONNXMatMulOpShapeHelper shapeHelper(&matMulOp);
+    LogicalResult shapecomputed = shapeHelper.computeShape(operandAdaptor);
+    assert(succeeded(shapecomputed) && "Could not compute output shape");
+
+    Type outputType = *op->result_type_begin();
+    assert(isRankedShapedType(outputType) && "Expected Ranked ShapedType");
+    ShapedType outputShapedType = outputType.cast<ShapedType>();
+    Type elementType = outputShapedType.getElementType();
+
+    Value A(operandAdaptor.A()), B(operandAdaptor.B());
+    auto aRank = A.getType().cast<ShapedType>().getShape().size();
+    auto bRank = B.getType().cast<ShapedType>().getShape().size();
+    // Size all the arrays to padded length.
+    int paddedRank = std::max(aRank, bRank);
+    paddedRank = std::max(paddedRank, 2);
+    DimsExpr aDims = shapeHelper.aDims;
+    DimsExpr bDims = shapeHelper.bDims;
+    llvm::BitVector aPadDims = shapeHelper.aPadDims;
+    llvm::BitVector bPadDims = shapeHelper.bPadDims;
+
+    DimsExpr outputDims = shapeHelper.dimsForOutput();
+    llvm::SmallVector<int64_t, 4> aShape;
+    llvm::SmallVector<int64_t, 4> bShape;
+    for (int64_t i = 0; i < paddedRank - 2; i++) {
+      aShape.push_back(outputDims[i].isLiteral() ? outputDims[i].getLiteral()
```

Comment:
suggestion: define a helper function to map an IndexExpr to its literal value or -1

Reply:
I'll add the helper function

Reply:
This function already exists, check in IndexExpr.hpp
```C++
  // Helpers for list of IndexExpressions
  static void getShape(llvm::SmallVectorImpl<IndexExpr> &indexExprList,
      llvm::SmallVectorImpl<int64_t> &intDimList);
```